paddle/fluid/operators/cumprod_op.cu

```diff
@@ -0,0 +1,365 @@
+e.g. x = [1, 4, 5, 0, 2, 3, 0];
+    zero_mask = [0, 0, 0, 1, 1, 1, 1];
+    zero_index = 3;
+when i < k, we need to calculate the result of sum{i<=j<n}(d_yj*yj),we can
+use reversed cumsum to calculate it.
+R = reversed_cumsum(dy[j]*y[j]);     //store the calculation result of the
+sum{i<=j<n}(d_yj*yj) and x[k+1],x[k+2],...,x[j] along the index k+1 ~ j.
+when i = k, we need to calculate the result of prod{i<w<j}(xw).
+To calculate it, we introduce x_filled_one, which fill 1 before x(i+1) along
+the index 0 ~ i.
+e.g. x = [1, 4, 5, 0, 2, 3, 0];
+    x_filled_one = [1, 1, 1, 1, 2, 3, 0];
+thus, we can use cumprod(x_filled_one[x[j]]) to calculate the result of
+prod{i<w<j}(xw).
+Then, we will show more detailed implement method.
+for (int i = 0; i < numel; i++) {
+    if (zero_mask[i] == 0) {       //case i < k
+        dx[i] = R[i] / x[i];
+        x_filled_one[i] = 1;
+    }
+    else {
+        if (i == 0) {              //case i = k
+            dx[i] = 1;
+            zero_index = i;
+            x_filled_one[i] = 1;
+        } else {
+            if (zero_mask[i-1] == 0) {    //case i = k
+                dx[i] = y[i-1];
+                zero_index = i;
+                x_filled_one[i] = 1;
+            }
+            else {                  //case i > k
+                dx[i] = 0;
+                x_filled_one[i] = x[i];
+            }
+        }
+    }
+}
+T = reversed_cumsum(dy[j]*cumprod(x_filled_one[x[j]]));
+if (zero_index != -1) {
+    dx[zero_index] *= T[zero_index];
+}
+*/
+
+template <typename T>
+class CumprodGradOpCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    const auto *x = ctx.Input<framework::Tensor>("X");
+    const auto *y = ctx.Input<framework::Tensor>("Out");
+    const auto *dy =
+        ctx.Input<framework::Tensor>(framework::GradVarName("Out"));
+    auto *dx = ctx.Output<framework::Tensor>(framework::GradVarName("X"));
+    auto dim = ctx.Attr<int>("dim");
+
+    size_t outer_dim, mid_dim, inner_dim;
+    GetCumprodDimInfo(x->dims(), dim, &outer_dim, &mid_dim, &inner_dim);
+    if (outer_dim == 0 || mid_dim == 0 || inner_dim == 0) return;
+
+    size_t numel = outer_dim * mid_dim * inner_dim;
+    size_t numel_x = x->numel();
+    size_t numel_y = y->numel();
+
+    const auto *x_data = x->data<T>();
+    const auto *y_data = y->data<T>();
+    const auto *dy_data = dy->data<T>();
+
+    auto place = BOOST_GET_CONST(platform::CUDAPlace, ctx.GetPlace());
+    const auto &dev_ctx =
+        ctx.template device_context<platform::CUDADeviceContext>();
+    auto *dx_data = dx->mutable_data<T>(place);
+
+    // deal with complex
+    const T *x_data_deal;
+    const T *y_data_deal;
+    memory::AllocationPtr x_conj;
+    memory::AllocationPtr y_conj;
+    if (framework::IsComplex<T>::value) {
+      x_conj = memory::Alloc(place, numel_x * sizeof(T));
+      auto *x_data_conj = reinterpret_cast<T *>(x_conj->ptr());
```

Comment:
Some suggestions. Why not just use `x_data_deal = ...` instead of `auto *x_data_conj = ...`? The `x_data_conj` here is not necessary. Same below.

Reply:
But I receive an error(invalid conversion from ‘const float*’ to ‘float*’) when I change as  mentioned above.

Reply:
The ConjFunctor is defined as ConjFunctor(const T* input, int64_t numel, T* output), so I need to use the temporary variables here.